src/client-side-metrics/operation-metrics-collector.ts

```diff
@@ -139,76 +209,92 @@ export class OperationMetricsCollector {
    * @param {grpc.status} attemptStatus The grpc status for the attempt.
    */
   onAttemptComplete(projectId: string, attemptStatus: grpc.status) {
-    if (
-      this.state ===
-        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET ||
-      this.state ===
-        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_SOME_ROWS_RECEIVED
-    ) {
-      this.state =
-        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS;
-      this.attemptCount++;
-      const endTime = new Date();
-      if (projectId && this.attemptStartTime) {
-        const totalTime = endTime.getTime() - this.attemptStartTime.getTime();
-        this.metricsHandlers.forEach(metricsHandler => {
-          if (metricsHandler.onAttemptComplete) {
-            metricsHandler.onAttemptComplete({
-              attemptLatency: totalTime,
-              serverLatency: this.serverTime ?? undefined,
-              connectivityErrorCount: this.connectivityErrorCount,
-              streaming: this.streamingOperation,
-              status: attemptStatus.toString(),
-              client_name: `nodejs-bigtable/${version}`,
-              metricsCollectorData: this.getMetricsCollectorData(),
-              projectId,
-            });
+    withMetricsDebug(() => {
+      checkState(
+        this.state,
+        [
+          MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET,
+          MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_SOME_ROWS_RECEIVED,
+        ],
+        () => {
+          this.state =
+            MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS;
+          this.attemptCount++;
+          const endTime = new Date();
+          if (projectId && this.attemptStartTime) {
+            const totalTime =
+              endTime.getTime() - this.attemptStartTime.getTime();
+            OperationMetricsCollector.metricsHandlers.forEach(
+              metricsHandler => {
+                if (metricsHandler.onAttemptComplete) {
+                  metricsHandler.onAttemptComplete({
+                    attemptLatency: totalTime,
+                    serverLatency: this.serverTime ?? undefined,
+                    connectivityErrorCount: this.connectivityErrorCount,
+                    streaming: this.streamingOperation,
+                    status: attemptStatus.toString(),
+                    client_name: `nodejs-bigtable/${version}`,
+                    metricsCollectorData: this.getMetricsCollectorData(),
+                    projectId,
+                  });
+                }
+              }
+            );
+          } else {
+            console.warn('ProjectId and start time should always be provided');
```

Comment:
should this be a throw?

It seems to me like all metrics errors should be handled with the METRICS_DEBUG logic. But maybe you wanted different levels for exceptions vs warnings?

Reply:
Yes. I think throwing here keeps the logic simpler so I made this change.